nodist/clang_check_attributes

```diff
@@ -0,0 +1,332 @@
+# licensed under Apache 2.0
+
+import argparse
+import sys
+import os
+from pathlib import Path
+from pprint import pprint
+from clang.cindex import CursorKind, Index
+
+
+# https://stackoverflow.com/a/74205044
+class FixSizedDict(dict):
+
+    def __init__(self, *args, maxlen=0, **kwargs):
+        self._maxlen = maxlen
+        super().__init__(*args, **kwargs)
+
+    def __setitem__(self, key, value):
+        dict.__setitem__(self, key, value)
+        if self._maxlen > 0:
+            if len(self) > self._maxlen:
+                self.pop(next(iter(self)))
+
+
+# by definition constructor is MT safe,
+# but functions that constructor call are not!
+MT_SAFE_KINDS = [CursorKind.CONSTRUCTOR]
+VERILATOR_ROOT = Path(__file__).parent.parent.resolve()
+IGNORE_FILES_LIST = [
+    # Dont process test files
+    "_test.cpp",
+    # Dont process Verilator coverage files
+    "Vlc",
+    # We want to use V3Const after astgen
+    "V3Const.cpp",
+    # Skip SystemC files
+    "_sc",
+    # Skip parser files
+    ".yy.cpp"
+]
+
+index = Index.create()
+# Cache of files already turned into translate unit
+# this speeds up checking when used function is from another
+# translation unit, but it increases memory consumption.
+# Limit max cache size to 20, to make sure we won't run OOM
+# in github actions
+translate_units = FixSizedDict(maxlen=20)
+# Already printed nodes in current function.
+# This prevents infinite loops due to recursion
+printed = []
+# Nodes that were marked as MT_SAFE, but are calling
+# function that isn't marked with this attribute
+unsafe_calls = []
+
+
+def get_diag_info(diag):
+    return {
+        'severity': diag.severity,
+        'location': diag.location,
+        'spelling': diag.spelling,
+        'ranges': list(diag.ranges),
+        'fixits': list(diag.fixits)
+    }
+
+
+def get_translate_unit(file, cfg):
+    file = str(file)
+    try:
+        translate_unit = translate_units[file]
+    except KeyError:
+        translate_unit = index.parse(file, cfg.clang_args)
+        if not translate_unit:
+            print(f"%Error: unable to load translate unit from file: {file}")
+            sys.exit(1)
+        for diag in translate_unit.diagnostics:
+            if diag.severity in [diag.Error, diag.Fatal]:
+                pprint(('diags',
+                        list(map(get_diag_info, translate_unit.diagnostics))))
+                sys.exit(1)
+        translate_units[file] = translate_unit
+    return translate_unit
+
+
+def print_node(node, level):
+    if fully_qualified_pretty(node) not in printed:
+        annotations = get_annotations(node)
+        filepath = Path(str(node.location.file)).resolve()
+        try:
+            filepath_relative = filepath.relative_to(VERILATOR_ROOT)
+            print('%s %s:%s \t\t %-35s %-100s %s' %
+                  (' ' * level, filepath_relative, node.extent.start.line,
+                   node.kind, fully_qualified_pretty(node), annotations))
```

Comment:
To appease editors, the filename needs to be on left, with colon after.  Also I don't think CursorKind adds useful information, from reading the method name it's pretty obvious (if you think is critical for debug then keep it but remove "CursorKind." prefix).  Also having attributes on far right adds a lot of spacing and generally IMO looks better to have all fixed indents on left.  Finally please don't use tabs.  So here's the first output I'd expect:

```
::group::VerilatedContext::commandArgs(int, char **)
include/verilated.h:418:             ['MT_SAFE']    VerilatedContext::commandArgs(int, char **)
include/verilated.cpp:2494:          []              VerilatedContext::commandArgs(int, const char **)
include/verilated.h:189:             ['MT_SAFE']      VerilatedLockGuard::VerilatedLockGuard(VerilatedMutex &)
include/verilated.h:163:             ['MT_SAFE']       VerilatedMutex::lock()
include/verilated.h:177:             ['MT_SAFE']        VerilatedMutex::try_lock()
include/verilated.h:536:             []               VerilatedContext::impp()
include/verilated_imp.h:390:         []               VerilatedContextImp::commandArgsAddGuts(int, const char **)
::endgroup::
```